### DISCUSSION TITLE
Seeking Guidance for Finetuning F5-TTS. Small datasets (10-60 minutes) on 12GB VRAM

### DISCUSSION THREAD

**Discussion by S-T-K** [SWivid/F5-TTS#769] [Q&A]
I've been getting great results with F5-TTS, but my first attempt at fine-tuning trained from scratch instead of using the pretrained model—the output started as noise and is only slowly becoming speech.

How do I correctly fine-tune instead of starting from scratch?

Do I need to set "Tokenizer File" and "Path to Pretrained Checkpoint" manually? If so, what should I put?
Does "Download corresponding dataset first, and fill in the path in scripts" ([repo link](https://github.com/SWivid/F5-TTS/tree/main/src/f5_tts/train)) refer to this?

**Project Details:**
I'm working on generating voices for characters from an old game. I have 10 to 60 minutes of clean audio samples per character. Language: English.

**Hardware:**
GPU: Nvidia 4080 Laptop (12GB VRAM)
I'm looking for advice on the best values to set for finetuning, given my hardware. Here’s what I’ve gathered so far, but I’d love some expert input:

**Parameter Questions**
Batch Size per GPU: I assume 6400 should work with 12GB VRAM, but would appreciate confirmation.
Max Samples: Not sure, but I read that 2 might be fine ([reference](https://github.com/SWivid/F5-TTS/discussions/57#discussioncomment-10959029)).
Gradient Accumulation Steps & Max Gradient Norm: No idea—should I just leave them at 1?
Epochs: How many would be reasonable for my dataset size?
Warmup Updates: Not sure what value is appropriate.
Save per Updates: I assume setting this high is better, as frequent saving would slow down training?
Last per Updates: Not sure what value to use here either.
Other Options:
Use 8-bit Adam optimizer – Should I enable this?
Mixed Precision – Any recommendations based on my GPU?
Logger – Not sure what’s best here.

**Finetuning Duration**
How long should I expect finetuning to take per character? Just so I can compare against my actual training times and check if my machine is underperforming due to driver or config issues.

Any guidance would be highly appreciated! 🚀

Thanks in advance!

**Comment by S-T-K** [comment]
I was able to answer some of the questions myself:
For finetuning, just check the checkbox of the same name. No need to specify anything else. 

On my machine any value for max samples higher than 8 leads to horribly slow training speed. So I guess going higher than that is not advisable.

Trained for 320 epochs but unfortunately the resulting model doesn't perform any better than the base model in my opinion. No discernable difference in the generated audio files.

**Reply by SWivid** [reply]
[try turn off use_ema and see if differences](https://github.com/SWivid/F5-TTS/tree/main/src/f5_tts/train#:~:text=The%20use_ema%20%3D%20True%20is%20harmful%20for%20early%2Dstage%20finetuned%20checkpoints%20(which%20goes%20just%20few%20updates%2C%20thus%20ema%20weights%20still%20dominated%20by%20pretrained%20ones)%2C%20try%20turn%20it%20off%20and%20see%20if%20provide%20better%20results.)

https://github.com/SWivid/F5-TTS/blob/261b2774f22caa4b621c4f115f69671385519c3f/src/f5_tts/infer/utils_infer.py#L231-L240

**Reply by S-T-K** [reply]
Thanks for the tip! Tried it, and I still can't hear any improvement. Sounds the same with use_ema on true or false. 

The generated audio already sounds very close to the reference sample even on the base model, so maybe that's why there is no noticeable difference?

**Reply by SWivid** [reply]
It's possible, there could be a slight improvement but not noticeable (TTS model has a relatively large marginal effect on timbre).

**Reply by S-T-K** [reply]
I see, that makes sense. Do you know of a metric or score I could calculate for the generated files to compare their similarity to the reference? For example, I could generate 100 files using the base model and the fine-tuned model and compare them statistically.

**Reply by SWivid** [reply]
https://github.com/SWivid/F5-TTS/tree/main/src/f5_tts/eval

**Reply by S-T-K** [reply]
Oh perfect, didn't know it's part of the repo. Thank you!

**Reply by SWivid** [reply]
Since you wanna evaluate for specific voice performance rather than zero-shot ability, you could write some code and do eval for the certain speaker voices, with https://github.com/SWivid/F5-TTS/blob/261b2774f22caa4b621c4f115f69671385519c3f/src/f5_tts/eval/utils_eval.py#L379-L413
and maybe also utmos

**Reply by S-T-K** [reply]
Update: Fine-tuning did improve the output quality significantly! 
Initially, I thought it hadn't, but that was due to an error I missed in the console. While comparing the original pretrained model and the fine-tuned model in the finetune gradio interface, I encountered an issue. When switching back to the original model, an error (not visible in the gradio interface) caused the fine-tuned model to remain loaded, even though I believed I had switched back. As a result, I was unknowingly comparing the fine-tuned model to itself, which is why I didn't notice much difference and mistakenly thought the original model also sounded good.

So in summary:  fine-tuning is worth it and f5-tts is amazing. Thank you so much @SWivid!

**Comment by Balauruu** [comment]
Hello, could you share the parameters used and what other insights you had with training such as the 8-bit Adam optimizer (if you found out how it affects your training). I am currently doing a very similar training with a similar setup and I'm getting some weird artefacts (a wet sound) and I'm not sure if the model is undertrained or my reference audio files are bad and it's picking too much background noise. Also on my machine (4070 laptop) it takes like 1 day for around 20k-25k steps depending on parameters and I can't figure out if it's performing normally or slow. (I see ppl here training for >100k steps). I would appreciate very much some insights and results from your training. 

My Parameters:
-learning_rate: 0.00001,
-batch_size_per_gpu: 2200 ( I found gives me the best training time per epoch)
-max_samples: 64,
-grad_accumulation_steps: 1,
-max_grad_norm: 1,
-epochs: 3425 ( I left it this high, because I thought it doesn't affect training. Just the slope in results when using the tensorboard to analyze?
-num_warmup_updates: 100
-save_per_updates: 2000,
-finetune: true,
-mixed_precision: "bf16",
"logger": "tensorboard",
- 8 bit optimizer: false (not enabled) (from what I researched online it should be a straight speed upgrade, but I saw people in other discussion boards are not using it when training)

 I actually trained my model for only 150 epochs, 30k steps.
